hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java

```diff
@@ -222,24 +222,38 @@ private CompactSerializableRegistration getOrCreateRegistration(Class clazz) {
     }
 
     private CompactSerializableRegistration getOrCreateRegistration(String typeName) {
-        return typeNameToRegistrationMap.computeIfAbsent(typeName, s -> {
-            Class<?> clazz;
-            try {
-                // When the registration does not exist, we treat typeName as className
-                // to check if there is a class with the given name in the classpath.
-                clazz = ClassLoaderUtil.loadClass(classLoader, typeName);
-            } catch (Exception e) {
-                // There is no such class that has typeName as its name.
-                // We should try to read this as GenericRecord. We are
-                // returning this registration here to remember that we
-                // should read instances of this typeName as GenericRecords,
-                // instead of trying to load a class with that name over
-                // and over.
-                return CompactSerializableRegistration.GENERIC_RECORD_REGISTRATION;
-            }
+        CompactSerializableRegistration currentRegistration = typeNameToRegistrationMap.get(typeName);
+        if (currentRegistration != null) {
+            return currentRegistration;
+        }
+        // Execute potentially long-lasting operation outside CHM lock in computeIfAbsent.
+        // Some special classloaders (eg. JetClassLoader) may try to access external resources
+        // and require other threads.
+        // We might try to load the same class multiple times in parallel but this is not a problem.
+        CompactSerializableRegistration newRegistration = getOrCreateRegistration0(typeName);
+
+        // Registration might have been created by a concurrent thread.
+        // If so, use that one instead.
+        return typeNameToRegistrationMap.computeIfAbsent(typeName, k -> newRegistration);
+    }
 
-            return getOrCreateRegistration(clazz);
-        });
+    private CompactSerializableRegistration getOrCreateRegistration0(String typeName) {
+        Class<?> clazz;
+        try {
+            // When the registration does not exist, we treat typeName as className
+            // to check if there is a class with the given name in the classpath.
+            clazz = ClassLoaderUtil.loadClass(classLoader, typeName);
+        } catch (Exception e) {
```

Comment:
What about other types of exceptions, e.g. some `IOException`, instead of `ClassNotFoundException`

Reply:
I only refactored the code without changing the logic. I do not have enough knowledge about compact serialization to determine if this logic is by design or problematic.
Your question is valid though.